Mathlib/Data/Fin/Tuple/Basic.lean

```diff
@@ -0,0 +1,878 @@
+    rw [succAbove_last]
+#align fin.insert_nth_last Fin.insert_nth_last
+
+@[simp]
+theorem insert_nth_last' (x : β) (p : Fin n → β) :
+    @insertNth _ (fun _ ↦ β) (last n) x p = snoc p x := by simp [insert_nth_last]
+#align fin.insert_nth_last' Fin.insert_nth_last'
+
+@[simp]
+theorem insert_nth_zero_right [∀ j, Zero (α j)] (i : Fin (n + 1)) (x : α i) :
+    i.insertNth x 0 = Pi.single i x :=
+  insert_nth_eq_iff.2 <| by simp [succAbove_ne, Pi.zero_def]
+#align fin.insert_nth_zero_right Fin.insert_nth_zero_right
+
+theorem insert_nth_binop (op : ∀ j, α j → α j → α j) (i : Fin (n + 1)) (x y : α i)
+    (p q : ∀ j, α (i.succAbove j)) :
+    (i.insertNth (op i x y) fun j ↦ op _ (p j) (q j)) = fun j ↦
+      op j (i.insertNth x p j) (i.insertNth y q j) :=
+  insert_nth_eq_iff.2 <| by simp
+#align fin.insert_nth_binop Fin.insert_nth_binop
+
+@[simp]
+theorem insert_nth_mul [∀ j, Mul (α j)] (i : Fin (n + 1)) (x y : α i)
+    (p q : ∀ j, α (i.succAbove j)) :
+    i.insertNth (x * y) (p * q) = i.insertNth x p * i.insertNth y q :=
+  insert_nth_binop (fun _ ↦ (· * ·)) i x y p q
+#align fin.insert_nth_mul Fin.insert_nth_mul
+
+@[simp]
+theorem insert_nth_add [∀ j, Add (α j)] (i : Fin (n + 1)) (x y : α i)
+    (p q : ∀ j, α (i.succAbove j)) :
+    i.insertNth (x + y) (p + q) = i.insertNth x p + i.insertNth y q :=
+  insert_nth_binop (fun _ ↦ (· + ·)) i x y p q
+#align fin.insert_nth_add Fin.insert_nth_add
+
+@[simp]
+theorem insert_nth_div [∀ j, Div (α j)] (i : Fin (n + 1)) (x y : α i)
+    (p q : ∀ j, α (i.succAbove j)) :
+    i.insertNth (x / y) (p / q) = i.insertNth x p / i.insertNth y q :=
+  insert_nth_binop (fun _ ↦ (· / ·)) i x y p q
+#align fin.insert_nth_div Fin.insert_nth_div
+
+@[simp]
+theorem insert_nth_sub [∀ j, Sub (α j)] (i : Fin (n + 1)) (x y : α i)
+    (p q : ∀ j, α (i.succAbove j)) :
+    i.insertNth (x - y) (p - q) = i.insertNth x p - i.insertNth y q :=
+  insert_nth_binop (fun _ ↦ Sub.sub) i x y p q
+#align fin.insert_nth_sub Fin.insert_nth_sub
+
+@[simp]
+theorem insert_nth_sub_same [∀ j, AddGroup (α j)] (i : Fin (n + 1)) (x y : α i)
+    (p : ∀ j, α (i.succAbove j)) : i.insertNth x p - i.insertNth y p = Pi.single i (x - y) := by
+  simp_rw [← insert_nth_sub, ← insert_nth_zero_right, Pi.sub_def, sub_self, Pi.zero_def]
+#align fin.insert_nth_sub_same Fin.insert_nth_sub_same
+
+variable [∀ i, Preorder (α i)]
+
+theorem insert_nth_le_iff {i : Fin (n + 1)} {x : α i} {p : ∀ j, α (i.succAbove j)} {q : ∀ j, α j} :
+    i.insertNth x p ≤ q ↔ x ≤ q i ∧ p ≤ fun j ↦ q (i.succAbove j) := by
+  simp [Pi.le_def, forall_iff_succ_above i]
+#align fin.insert_nth_le_iff Fin.insert_nth_le_iff
+
+theorem le_insert_nth_iff {i : Fin (n + 1)} {x : α i} {p : ∀ j, α (i.succAbove j)} {q : ∀ j, α j} :
+    q ≤ i.insertNth x p ↔ q i ≤ x ∧ (fun j ↦ q (i.succAbove j)) ≤ p := by
+  simp [Pi.le_def, forall_iff_succ_above i]
+#align fin.le_insert_nth_iff Fin.le_insert_nth_iff
+
+open Set
+
+-- Porting note: Had to up maxHeartbeats
+set_option maxHeartbeats 1000000
```

Comment:
Can we limit this to only the declaration that needs it?

Reply:
No longer needed